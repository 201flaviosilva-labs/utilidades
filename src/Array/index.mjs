import { allEqual } from "./allEqual.mjs";
import { choice } from "./choice.mjs";
import { findBigObject } from "./findBigObject.mjs";
import { findLowObject } from "./findLowObject.mjs";
import { isSorted } from "./isSorted.mjs";
import { moveLeft, moveRight } from "./Move.mjs";
import { shuffle } from "./shuffle.mjs";
import { sortAscending } from "./sortAscending.mjs";
import { sortAscendingObject } from "./sortAscendingObject.mjs";
import { sortDescending } from "./sortDescending.mjs";
import { sortDescendingObject } from "./sortDescendingObject.mjs";

/**
 * Functions utils for arrays
 * 
 * @namespace Array
 */
export {
	allEqual,
	choice,
	findBigObject,
	findLowObject,
	isSorted,
	moveLeft, moveRight,
	shuffle,
	sortAscending, sortAscendingObject,
	sortDescending, sortDescendingObject,
};
