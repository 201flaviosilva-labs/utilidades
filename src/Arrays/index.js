import { allEqual } from "./allEqual.js";
import { choice } from "./choice.js";
import { findBigObject } from "./findBigObject.js";
import { findLowObject } from "./findLowObject.js";
import { isSorted } from "./isSorted.js";
import { moveLeft, moveRight } from "./move.js";
import { shuffle } from "./shuffle.js";
import { sortAscending } from "./sortAscending.js";
import { sortAscendingObject } from "./sortAscendingObject.js";
import { sortDescending } from "./sortDescending.js";
import { sortDescendingObject } from "./sortDescendingObject.js";

/**
 * Functions utils for arrays
 * 
 * @namespace Arrays
 */
export {
	allEqual,
	choice,
	findBigObject,
	findLowObject,
	isSorted,
	moveLeft, moveRight,
	shuffle,
	sortAscending, sortAscendingObject,
	sortDescending, sortDescendingObject,
};
