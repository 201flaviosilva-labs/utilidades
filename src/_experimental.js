// Points
export function distanceTwoPoints(x1, y1, x2, y2) { return Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2)); }
export function distanceTwoPointsVector2(point1, point2) { return distanceTwoPoints(point1.x, point1.y, point2.x, point2.y); }

export function angleBetweenTwoPoints(x1, y1, x2, y2) { return Math.atan2(y2 - y1, x2 - x1); }
export function angleBetweenTwoPointsVector2(point1, point2) { return angleBetweenTwoPoints(point1.x, point1.y, point2.x, point2.y); }
export function angleBetweenTwoPointsDegrees(x1, y1, x2, y2) { return angleBetweenTwoPointsRadians(x1, y1, x2, y2) * 180 / Math.PI; }
export function angleBetweenTwoPointsVector2Degrees(point1, point2) { return angleBetweenTwoPointsDegrees(point1.x, point1.y, point2.x, point2.y); }

// Circles
export function distanceBetweenTwoCirclesFromCircle(circle1, circle2) {
	return distanceTwoPoints(circle1.x, circle1.y, circle2.x, circle2.y) - (circle1.radius + circle2.radius);
}
export function distanceBetweenTwoCircles(x1, y1, radius1, x2, y2, radius2) {
	return distanceBetweenTwoCirclesFromCircle(
		{ x: x1, y: y1, radius: radius1 },
		{ x: x2, y: y2, radius: radius2 }
	);
}

export function circleArea(radius) { return Math.PI * radius * radius; }
export function circlePerimeter(radius) { return 2 * Math.PI * radius; }

// Rectangles
export function distanceBetweenTwoRectanglesFromBounds(rect1, rect2) {
	const xOverlap = Math.min(rect1.x + rect1.width, rect2.x + rect2.width) - Math.max(rect1.x, rect2.x);
	const yOverlap = Math.min(rect1.y + rect1.height, rect2.y + rect2.height) - Math.max(rect1.y, rect2.y);
	return xOverlap * yOverlap;
}
export function distanceBetweenTwoRectangles(x1, y1, width1, height1, x2, y2, width2, height2) {
	return distanceBetweenTwoRectanglesFromBounds(
		{ x: x1 + width1 / 2, y: y1 + height1 / 2, width: width1, height: height1 },
		{ x: x2 + width2 / 2, y: y2 + height2 / 2, width: width2, height: height2 }
	);
}

export function angleBetweenTwoRectanglesFromBounds(rect1, rect2) {
	const xOverlap = Math.min(rect1.x + rect1.width, rect2.x + rect2.width) - Math.max(rect1.x, rect2.x);
	const yOverlap = Math.min(rect1.y + rect1.height, rect2.y + rect2.height) - Math.max(rect1.y, rect2.y);
	return Math.atan2(yOverlap, xOverlap);
}
export function angleBetweenTwoRectangles(x1, y1, width1, height1, x2, y2, width2, height2) {
	return angleBetweenTwoRectanglesFromBounds(
		{ x: x1 + width1 / 2, y: y1 + height1 / 2, width: width1, height: height1 },
		{ x: x2 + width2 / 2, y: y2 + height2 / 2, width: width2, height: height2 }
	);
}

export function rectangleArea(width, height) { return width * height; }
export function rectanglePerimeter(width, height) { return 2 * (width + height); }
export function rectanglePerimeterFromBounds(rectangle) { return rectanglePerimeter(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }

export function rectangleVertices(x, y, width, height) {
	return [
		{ x: x, y: y },
		{ x: x + width, y: y },
		{ x: x + width, y: y + height },
		{ x: x, y: y + height }
	];
}
export function rectangleVerticesFromBounds(rectangle) { return rectangleVertices(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }

export function rectangleCenterX(x, width) { return x + width / 2; }
export function rectangleCenterY(y, height) { return y + height / 2; }
export function rectangleCenter(x, y, width, height) { return { x: rectangleCenterX(x, width), y: rectangleCenterY(y, height) }; }
export function rectangleCenterXFromBounds(rectangle) { return rectangleCenterX(rectangle.x, rectangle.width); }
export function rectangleCenterYFromBounds(rectangle) { return rectangleCenterY(rectangle.y, rectangle.height); }
export function rectangleCenterFromBounds(rectangle) { return rectangleCenter(rectangle.x, rectangle.y, rectangle.width, rectangle.height); }
