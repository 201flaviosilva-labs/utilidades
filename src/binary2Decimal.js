/**
 * Convert a binary number to a decimal
 * 
 * @example
 * binary2Decimal("101010"); // 42
 * 
 * @param {string} binary - the value to be converted
 * @returns {number} - conversion decimal
 */
export function binary2Decimal(binary) {
	if (typeof binary === "string") return parseInt(binary, 2);
}
