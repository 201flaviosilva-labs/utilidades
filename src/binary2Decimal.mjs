/**
 * Convert a number/string to a decimal
 * 
 * @example
 * binary2Decimal(101010); // 42
 * 
 * @param {number|string} binary - the number/string to be converted
 * @returns {number} - conversion decimal
 */
export function binary2Decimal(binary) {
	if (typeof binary === "string") return parseInt(binary.split("").reverse().join(""), 2);
	else if (typeof binary === "number") return parseInt(binary, 2);
}
