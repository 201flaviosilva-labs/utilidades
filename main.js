import * as utilidades from "./src/main";

console.log(utilidades);

document.querySelector("#app").innerHTML = `
  <div>
    <a href="https://vitejs.dev" target="_blank">
      <img src="/vite.svg" class="logo" alt="Vite logo" />
    </a>
    <h1>Utilidades Testes</h1>
  </div>
`;
