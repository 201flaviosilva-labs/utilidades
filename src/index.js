import * as Arrays from "./Arrays/index.js";
import * as Maths from "./Maths/index.js";
import * as Physics from "./Physics/index.js";

import { allCharactersSame } from "./allCharactersSame.js";
import { binary2Decimal } from "./binary2Decimal.js";
import { BinarySearchTree, BinarySearchTreeInstance } from "./BinarySearchTree.js";
import { choiceTrend } from "./choiceTrend.js";
import { clone } from "./clone.js";
import { compare2Objects } from "./compare2Objects.js";
import { decimal2Binary } from "./decimal2Binary.js";
import { getDate } from "./getDate.js"; // getTime, getMilliseconds, getSeconds, getMinutes, getHours, getDay, getWeekDay, getWeek, getMonth, getYear, getDateFormatted 
import { EventSystem, EventSystemInstance } from "./EventSystem.js";
import { Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci } from "./Fibonacci.js";
import { getVersion } from "./getVersion.js";
import { isObjectEmpty } from "./isObjectEmpty.js";
import { isFalsy, isTruthy } from "./isTruthyFalsy.js";
import { and, or, xor } from "./logicalOperators.js";
import { randomColor, randomColor0X, randomRGBColor, randomRGBAColor } from "./randomColor.js";
import { randomNumber, randomInt, randomFloat } from "./randomNumber.js";
import { randomString } from "./randomString.js";
import { randomWalk1D, randomWalk2D, randomWalk3D } from "./randomWalk.js";
import { reverseString } from "./reverseString.js";
import { Vector2 } from "./Vector2.js";

export {
	Arrays,
	Maths,
	Physics,

	BinarySearchTree, BinarySearchTreeInstance,
	EventSystem, EventSystemInstance,
	Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci,
	Vector2,

	allCharactersSame,
	binary2Decimal,
	compare2Objects,
	decimal2Binary,
	choiceTrend,
	clone,
	getDate,
	getVersion,
	isObjectEmpty,
	isFalsy, isTruthy,
	and, or, xor, // logicalOperators
	randomColor, randomColor0X, randomRGBColor, randomRGBAColor,
	randomNumber, randomInt, randomFloat,
	randomString,
	randomWalk1D, randomWalk2D, randomWalk3D,
	reverseString,
};
