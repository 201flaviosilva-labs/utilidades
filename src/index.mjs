import * as Array from "./Array/index.mjs";
// import * as DOM from "./DOM/index.mjs";
import * as Maths from "./Maths/index.mjs";
// import * as SortingAlgorithms from "./SortingAlgorithms/index.mjs";

import { allCharactersSame } from "./allCharactersSame.mjs";
import { binary2Decimal } from "./binary2Decimal.mjs";
import { BinarySearchTree, BinarySearchTreeInstance } from "./BinarySearchTree.mjs";
import { stringToCamelCase, camelCaseToNormal } from "./camelCase.mjs";
import { stringToCapitalize, capitalizeToNormal } from "./capitalizeCase.mjs";
import { choiceTrend } from "./choiceTrend.mjs";
import { clone } from "./clone.mjs";
import { compare2Objects } from "./compare2Objects.mjs";
import { decimal2Binary } from "./decimal2Binary.mjs";
import { getDate } from "./getDate.mjs"; // getTime, getMilliseconds, getSeconds, getMinutes, getHours, getDay, getWeekDay, getWeek, getMonth, getYear, getDateFormatted 
import { EventSystem, EventSystemInstance } from "./EventSystem.mjs";
import { Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci } from "./Fibonacci.mjs";
import { FIFO } from "./FIFO.mjs";
// import { getUrlParameter } from "./getUrlParameter.mjs";
// import { getVersion } from "./getVersion.mjs";
import { invertSentence, invertWords } from "./invertText.mjs";
import { isFalsy, isTruthy } from "./isTruthyFalsy.mjs";
import { stringToKebabCase, kebabCaseToNormal } from "./kebabCase.mjs";
import { and, or, xor } from "./logicalOperators.mjs";
import { LIFO } from "./LIFO.mjs";
import { randomColor, randomColor0X, randomRGBColor, randomRGBAColor } from "./randomColor.mjs";
import { randomNumber, randomInt, randomFloat } from "./randomNumber.mjs";
import { randomString } from "./randomString.mjs";
import { randomWalk1D, randomWalk2D, randomWalk3D } from "./randomWalk.mjs";
import { stringToScreamingSnakeCase, screamingSnakeCaseToNormal } from "./screamingSnakeCase.mjs";
import { stringToSnakeCase, snakeCaseToNormal } from "./snakeCase.mjs";
import { topDownCarMovimentation } from "./topDownCarMovimentation.mjs";
import { Vector2 } from "./Vector2.mjs";

export {
	Array,
	// DOM,
	Maths,
	// SortingAlgorithms,

	BinarySearchTree, BinarySearchTreeInstance,
	EventSystem, EventSystemInstance,
	Fibonacci, fibonacciSequence, fibonacciUntil, fibonacciCustomSequence, recursiveFibonacci,
	FIFO,
	LIFO,
	Vector2,

	allCharactersSame,
	binary2Decimal,
	stringToCamelCase, camelCaseToNormal,
	stringToCapitalize, capitalizeToNormal,
	choiceTrend,
	clone,
	compare2Objects,
	decimal2Binary,
	getDate,
	// getUrlParameter,
	// getVersion,
	invertSentence, invertWords,
	isFalsy, isTruthy,
	stringToKebabCase, kebabCaseToNormal,
	and, or, xor, // logicalOperators
	randomColor, randomColor0X, randomRGBColor, randomRGBAColor,
	randomNumber, randomInt, randomFloat,
	randomString,
	randomWalk1D, randomWalk2D, randomWalk3D,
	stringToScreamingSnakeCase, screamingSnakeCaseToNormal,
	stringToSnakeCase, snakeCaseToNormal,
	topDownCarMovimentation,
};
