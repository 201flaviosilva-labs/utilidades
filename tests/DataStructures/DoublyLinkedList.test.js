import { expect, it } from "vitest";
import { DataStructures } from "./src/main";
const { DoublyLinkedList, DLLNode: Node } = DataStructures;

describe("DataStructures/DoublyLinkedList.js", () => {
	describe("DoublyLinkedList", () => {
		describe("constructor()", () => {
			it("creates a new DoublyLinkedList, starting with a value", () => {
				const dll = new DoublyLinkedList(100);

				expect(dll.head).toBeInstanceOf(Node);
				expect(dll.head.value).toBe(100);
				expect(dll.head.next).toBeNull();
				expect(dll.head.prev).toBeNull();

				expect(dll.tail).toBeInstanceOf(Node);
				expect(dll.tail.value).toBe(100);
				expect(dll.tail.next).toBeNull();
				expect(dll.tail.prev).toBeNull();

				expect(dll.size).toBe(1);
				expect(dll.toArray()).toEqual([100]);
				expect(dll).toMatchSnapshot();
			});

			it("creates a new DoublyLinkedList, starting with an array", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.head.value).toBe(10);
				expect(dll.tail.value).toBe(30);

				expect(dll.size).toBe(3);
				expect(dll.toArray()).toEqual([10, 20, 30]);
				expect(dll).toMatchSnapshot();
			});

			it("creates a new string DoublyLinkedList", () => {
				const dll = new DoublyLinkedList(["A", "B", "C"]);

				expect(dll.head.value).toBe("A");
				expect(dll.tail.value).toBe("C");

				expect(dll.size).toBe(3);
				expect(dll.toArray()).toEqual(["A", "B", "C"]);
			});

			it("creates a new empty DoublyLinkedList", () => {
				const dll = new DoublyLinkedList();
				expect(dll.size).toBe(0);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.toArray()).toEqual([]);
				expect(dll).toMatchSnapshot();
			});

			it("creates a new empty DoublyLinkedList, starting with an empty array", () => {
				const dll = new DoublyLinkedList([]);
				expect(dll.size).toBe(0);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.toArray()).toEqual([]);
				expect(dll).toMatchSnapshot();
			});
		});

		describe("clear()", () => {
			it("should clear the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				dll.clear();
				expect(dll.size).toBe(0);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.toArray()).toEqual([]);
			});

			it("should clear an empty list", () => {
				const dll = new DoublyLinkedList();
				dll.clear();
				expect(dll.size).toBe(0);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.toArray()).toEqual([]);
			});
		});

		describe("get()", () => {
			it("should returns the value of the node at the given index", () => {
				const dll = new DoublyLinkedList([10, 20, 30, 40, 50]);
				expect(dll.get(0)).toBe(10);
				expect(dll.get(1)).toBe(20);
				expect(dll.get(2)).toBe(30);
				expect(dll.get(3)).toBe(40);
				expect(dll.get(4)).toBe(50);
			});

			it("should returns the node at the given index", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				const node = dll.get(1, true);

				expect(node.value).toBe(20);
				expect(node.prev.value).toBe(10);
				expect(node.next.value).toBe(30);
				expect(node).toBeInstanceOf(Node);
			});

			it("should return undefined if the index is out of bounds", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.get(3)).toBeUndefined();
				expect(dll.get(-1)).toBeUndefined();
			});

			it("should return undefined if the list is empty", () => {
				const dll = new DoublyLinkedList();
				expect(dll.get(0)).toBeUndefined();
			});
		});

		describe("set()", () => {
			it("should change a node value by the given index and new value", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.set(0, 100)).toBe(true);
				expect(dll.set(1, 200)).toBe(true);
				expect(dll.set(2, 300)).toBe(true);

				expect(dll.toArray()).toEqual([100, 200, 300]);
				expect(dll).toMatchSnapshot();
			});

			it("should return false if the given index is out of bounds", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.set(3, 500)).toBe(false);
				expect(dll.set(-1, 500)).toBe(false);
			});

			it("should return false if the list is empty", () => {
				const dll = new DoublyLinkedList();
				expect(dll.set(0, 500)).toBe(false);
			});
		});

		describe("unshift()", () => {
			it("adds a new value to the start of the list", () => {
				const dll = new DoublyLinkedList([20, 30]);
				expect(dll.size).toBe(2);

				dll.unshift(10);

				expect(dll.head.value).toBe(10);
				expect(dll.head.next.value).toBe(20);
				expect(dll.head.prev).toBeNull();
				expect(dll.tail.value).toBe(30);

				expect(dll.size).toBe(3);
				expect(dll.toArray()).toEqual([10, 20, 30]);
				expect(dll).toMatchSnapshot();
			});

			it("should add a new value to the start of an empty list", () => {
				const dll = new DoublyLinkedList();
				expect(dll.size).toBe(0);

				dll.unshift(10);

				expect(dll.head.value).toBe(10);
				expect(dll.head.next).toBeNull();
				expect(dll.head.prev).toBeNull();

				expect(dll.size).toBe(1);
				expect(dll.toArray()).toEqual([10]);
				expect(dll).toMatchSnapshot();
			});
		});

		describe("push()", () => {
			it("should add a new value to the end of list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.push(40)).toBeTruthy();
				expect(dll.tail.value).toBe(40);
				expect(dll.toArray()).toEqual([10, 20, 30, 40]);
			});

			it("should add a new value to the end of an empty list", () => {
				const dll = new DoublyLinkedList();

				expect(dll.push(40)).toBeTruthy();
				expect(dll.head.value).toBe(40);
				expect(dll.tail.value).toBe(40);
				expect(dll.toArray()).toEqual([40]);
			});
		});

		describe("insert()", () => {
			it("should insert a new value at the given index", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.insert(1, 40)).toBeInstanceOf(DoublyLinkedList);

				expect(dll.size).toBe(4);
				expect(dll.toArray()).toEqual([10, 40, 20, 30]);
			});

			it("should insert a new value at the start of the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				dll.insert(0, 40);

				expect(dll.size).toBe(4);
				expect(dll.toArray()).toEqual([40, 10, 20, 30]);
			});

			it("should insert a new value at the end of the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				dll.insert(3, 40);

				expect(dll.size).toBe(4);
				expect(dll.toArray()).toEqual([10, 20, 30, 40]);
			});

			it("should insert a new in a empty list", () => {
				const dll = new DoublyLinkedList();
				dll.insert(0, 10);

				expect(dll.size).toBe(1);
				expect(dll.toArray()).toEqual([10]);
			});

			it("should return false if the given index is out of bounds", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.insert(4, 500)).toBe(false);
				expect(dll.insert(-1, 500)).toBe(false);
				expect(dll.size).toBe(3);
			});
		});

		describe("shift()", () => {
			it("should remove the first value from the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.shift()).toBe(10);
				expect(dll.head.value).toEqual(20);
				expect(dll.head.prev).toBeNull();
				expect(dll.head.next.value).toEqual(30);
				expect(dll.tail.value).toEqual(30);

				expect(dll.size).toBe(2);
				expect(dll.toArray()).toEqual([20, 30]);
				expect(dll).toMatchSnapshot();
			});

			it("should clear the list if it's the last node", () => {
				const dll = new DoublyLinkedList(10);

				expect(dll.shift()).toBe(10);
				expect(dll.toArray()).toEqual([]);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
			});

			it("should return undefined if the list is empty", () => {
				const dll = new DoublyLinkedList();
				expect(dll.shift()).toBeUndefined();
				expect(dll.toArray()).toEqual([]);
			});

			it("should remove all the values from the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.shift()).toBe(10);
				expect(dll.shift()).toBe(20);
				expect(dll.shift()).toBe(30);

				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.size).toBe(0);
				expect(dll.toArray()).toEqual([]);
				expect(dll).toMatchSnapshot();
			});
		});

		describe("pop()", () => {
			it("should remove the last value from the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.pop()).toBe(30);
				expect(dll.tail.value).toEqual(20);
				expect(dll.tail.next).toBeNull();
				expect(dll.tail.prev.value).toEqual(10);

				expect(dll.size).toBe(2);
				expect(dll.toArray()).toEqual([10, 20]);
				expect(dll).toMatchSnapshot();
			});

			it("should clear the list if it's the last node", () => {
				const dll = new DoublyLinkedList(10);

				expect(dll.pop()).toBe(10);
				expect(dll.toArray()).toEqual([]);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
			});

			it("should return undefined if the list is empty", () => {
				const dll = new DoublyLinkedList();
				expect(dll.pop()).toBeUndefined();
				expect(dll.toArray()).toEqual([]);
			});

			it("should remove all the values from the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.pop()).toBe(30);
				expect(dll.pop()).toBe(20);
				expect(dll.pop()).toBe(10);

				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
				expect(dll.size).toBe(0);
				expect(dll.toArray()).toEqual([]);
				expect(dll).toMatchSnapshot();
			});
		});

		describe("remove()", () => {
			it("should remove the node at the given index", () => {
				const dll = new DoublyLinkedList([10, 20, 30, 40, 50]);

				expect(dll.remove(2)).toBe(30);
				expect(dll.size).toBe(4);
				expect(dll.toArray()).toEqual([10, 20, 40, 50]);
				expect(dll).toMatchSnapshot();
			});

			it("should remove the first node if the index is 0", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.remove(0)).toBe(10);
				expect(dll.size).toBe(2);
				expect(dll.toArray()).toEqual([20, 30]);
			});

			it("should remove the last node if the index is the length of the list", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.remove(2)).toBe(30);
				expect(dll.size).toBe(2);
				expect(dll.toArray()).toEqual([10, 20]);
			});


			it("should clear the list if it's the last node", () => {
				const dll = new DoublyLinkedList(10);

				expect(dll.remove(0)).toBe(10);
				expect(dll.toArray()).toEqual([]);
				expect(dll.head).toBeNull();
				expect(dll.tail).toBeNull();
			});

			it("should return undefined if the index is out of bounds", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);

				expect(dll.remove(3)).toBe(false);
				expect(dll.size).toBe(3);
				expect(dll.toArray()).toEqual([10, 20, 30]);
			});
		});

		describe("toArray()", () => {
			it("should convert the list to an array", () => {
				const dll = new DoublyLinkedList([10, 20, 30]);
				expect(dll.toArray()).toEqual([10, 20, 30]);
			});

			it("should return an empty array if the list is empty", () => {
				const dll = new DoublyLinkedList();
				expect(dll.toArray()).toEqual([]);
			});
		});
	});

	describe("Node", () => {
		it("creates a new Node with a value", () => {
			const node = new Node(10);

			expect(node.value).toBe(10);
			expect(node.next).toBeNull();
			expect(node.prev).toBeNull();
			expect(node).toMatchSnapshot();
		});

		it("should return a error if the value is not provided", () => {
			expect(() => new Node()).toThrowError("Node value cannot be undefined.");
		});
	});
});
