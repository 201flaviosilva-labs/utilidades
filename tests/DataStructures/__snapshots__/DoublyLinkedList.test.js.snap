// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > constructor() > creates a new DoublyLinkedList, starting with a value 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "prev": null,
    "value": 100,
  },
  "size": 1,
  "tail": Node {
    "next": null,
    "prev": null,
    "value": 100,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > constructor() > creates a new DoublyLinkedList, starting with an array 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "prev": [Circular],
        "value": 30,
      },
      "prev": [Circular],
      "value": 20,
    },
    "prev": null,
    "value": 10,
  },
  "size": 3,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": Node {
        "next": [Circular],
        "prev": null,
        "value": 10,
      },
      "value": 20,
    },
    "value": 30,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > constructor() > creates a new empty DoublyLinkedList 1`] = `
DoublyLinkedList {
  "head": null,
  "size": 0,
  "tail": null,
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > constructor() > creates a new empty DoublyLinkedList, starting with an empty array 1`] = `
DoublyLinkedList {
  "head": null,
  "size": 0,
  "tail": null,
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > pop() > should remove all the values from the list 1`] = `
DoublyLinkedList {
  "head": null,
  "size": 0,
  "tail": null,
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > pop() > should remove the last value from the list 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": null,
      "prev": [Circular],
      "value": 20,
    },
    "prev": null,
    "value": 10,
  },
  "size": 2,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": null,
      "value": 10,
    },
    "value": 20,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > remove() > should remove the node at the given index 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": Node {
          "next": null,
          "prev": [Circular],
          "value": 50,
        },
        "prev": [Circular],
        "value": 40,
      },
      "prev": [Circular],
      "value": 20,
    },
    "prev": null,
    "value": 10,
  },
  "size": 4,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": Node {
        "next": [Circular],
        "prev": Node {
          "next": [Circular],
          "prev": null,
          "value": 10,
        },
        "value": 20,
      },
      "value": 40,
    },
    "value": 50,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > set() > should change a node value by the given index and new value 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "prev": [Circular],
        "value": 300,
      },
      "prev": [Circular],
      "value": 200,
    },
    "prev": null,
    "value": 100,
  },
  "size": 3,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": Node {
        "next": [Circular],
        "prev": null,
        "value": 100,
      },
      "value": 200,
    },
    "value": 300,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > shift() > should remove all the values from the list 1`] = `
DoublyLinkedList {
  "head": null,
  "size": 0,
  "tail": null,
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > shift() > should remove the first value from the list 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": null,
      "prev": [Circular],
      "value": 30,
    },
    "prev": null,
    "value": 20,
  },
  "size": 2,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": null,
      "value": 20,
    },
    "value": 30,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > unshift() > adds a new value to the start of the list 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "prev": [Circular],
        "value": 30,
      },
      "prev": [Circular],
      "value": 20,
    },
    "prev": null,
    "value": 10,
  },
  "size": 3,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": Node {
        "next": [Circular],
        "prev": null,
        "value": 10,
      },
      "value": 20,
    },
    "value": 30,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > DoublyLinkedList > unshift() > should add a new value to the start of an empty list 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "prev": null,
    "value": 10,
  },
  "size": 1,
  "tail": Node {
    "next": null,
    "prev": null,
    "value": 10,
  },
}
`;

exports[`DataStructures/DoublyLinkedList.js > Node > creates a new Node with a value 1`] = `
Node {
  "next": null,
  "prev": null,
  "value": 10,
}
`;
